feat: make docs reflecting latest nullability changes. (#586)
vaadin/docs: articles/fusion/advanced/endpoints-generator.asciidoc

* `org.eclipse.jdt.annotation.NonNull`
* any other annotation (including custom) that has the `nonnull` (case-insensitive) name.

==== Endpoint Functions

> For the endpoint function, two things can be nullable:

Reviewer: “Nullable” implies the ability, “can be” feels redundant.

```suggestion
For the endpoint function, two things are nullable:
```

* **Function parameter type**.

> It can be nullable, but it cannot be omitted.

Reviewer: A bit mixing the parameters (the receiving variables) and the arguments (values that are being send) here, let’s clarify.

```suggestion
Arguments cannot be omitted even when the parameter types are nullable.
```

> If you have that parameter and want it to be `null` in Java, you must send an `undefined` there during the function call.

Reviewer: ```suggestion
To receive a `null` parameter value in Java, send an `undefined` argument in the endpoint function call.
```

* **Function return type**.

.Original Java endpoint class
[source,java]
----
@Endpoint
class PersonEndpoint {
    // Person must have at least the first and the last name
    public void setFullName(@Nonnull String firstName, @Nonnull String lastName, String middleName) {
        // omitted code
    }

    // Full name must exist
    @Nonnull
    public String getFullName() {
        // omitted code
    }

    // Person can have no connections with other people. But if they have,
    // the connection cannot be null.
    public Map<String, @Nonnull String> getConnections() {
        // omitted code
    }
}
----

.Generated TypeScript endpoint functions
[source,typescript]
----
export async function setName(
  firstName: string,
  lastName: string,
  middleName: string | undefined
) {
  return client.call('PersonEndpoint', 'setFullName', {firstName, lastName, middleName});
}

export async function getFullName(): Promise<string> {
  return client.call('PersonEndpoint', 'getFullName');
}

export async function getConnections(): Promise<Record<string, string> | undefined> {
  return client.call('PersonEndpoint', 'getConnections');
}
----

==== Data Class Properties

> For the data class, one thing can be nullable:

Reviewer: For me this literal “one thing” mention and the list feel redundant. I’d suggest to remove the list, and only keep the paragraph that says that class properties are nullable.

> For the data class, nullable elements are:

Reviewer: Looks redundant to have a list with just one item, I’d remove the list and describe the item right away.

```suggestion
Properties of the data classes are nullable.
```

> * **Properties**.

Reviewer: ```suggestion
```

* `org.eclipse.jdt.annotation.NonNull`
* any other annotation (including custom) that has the `nonnull` (case-insensitive) name.

==== Endpoint Functions

> For the endpoint function, there are two things that can be nullable:

Reviewer 2: two things can be nullable

* **Function parameter type**.

> It can be nullable, but it cannot be omitted. If you have that parameter and want it to be `null` in Java, you must send an `undefined` there during the function call.

Reviewer 2: one sentence per line, i.e. put If you ... in a new line.

* **Function return type**.

.Original Java endpoint class
[source,java]
----
@Endpoint
class PersonEndpoint {
    // Person must have at least the first and the last name
    public void setFullName(@Nonnull String firstName, @Nonnull String lastName, String middleName) {
        // omitted code
    }

    // Full name must exist
    @Nonnull
    public String getFullName() {
        // omitted code
    }

    // Person can have no connections with other people. But if they have,
    // the connection cannot be null.
    public Map<String, @Nonnull String> getConnections() {
        // omitted code
    }
}
----

.Generated TypeScript endpoint functions
[source,typescript]
----
export async function setName(
  firstName: string,
  lastName: string,
  middleName: string | undefined
) {
  return client.call('PersonEndpoint', 'setFullName', {firstName, lastName, middleName});
}

export async function getFullName(): Promise<string> {
  return client.call('PersonEndpoint', 'getFullName');
}

export async function getConnections(): Promise<Record<string, string> | undefined> {
  return client.call('PersonEndpoint', 'getConnections');
}
----

==== Data Class Properties

> For the data class, there are one thing that can be nullable:

Reviewer 2: For the data class, one thing can be nullable:

}
----

==== Collection Item Types

> For the collection types, one thing can be nullable:

Reviewer: Also redundant “one thing” list here.

> For the collection types, there are one thing that can be nullable:

Reviewer 2: For the collection types, one thing can be nullable:

> For collection types, nullable elements are:
> 
> * **Item type**.

Reviewer: ```suggestion
The collection item type is nullable.
```
